warehouse/schema_test.go

```diff
@@ -220,4 +221,226 @@ var _ = Describe("Schema", func() {
 			})
 		})
 	})
+
+	Describe("Excluded schema", func() {
+		uploadSchema := warehouseutils.SchemaT{
+			"demo_event": {
+				"anonymous_id":       "string",
+				"channel":            "string",
+				"context_ip":         "string",
+				"context_passed_ip":  "string",
+				"context_request_ip": "string",
+				"event_text":         "string",
+				"group_id":           "string",
+				"id":                 "string",
+				"loaded_at":          "datetime",
+				"name":               "string",
+				"original_timestamp": "datetime",
+				"previous_id":        "string",
+				"received_at":        "datetime",
+				"record_id":          "string",
+				"sent_at":            "datetime",
+				"timestamp":          "datetime",
+				"user_id":            "string",
+				"uuid_ts":            "datetime", // end of 18 rudder columns
+				"ad_name":            "string",
+				"blendo_id":          "string",
+				"clicks":             "int",
+				"currency":           "float",
+				"date_start":         "datatime",
+				"sample_event":       "json",
+			},
+		}
+
+		schemaInWarehouse := warehouseutils.SchemaT{
+			"demo_event": {
+				"anonymous_id":       "string",
+				"channel":            "string",
+				"context_ip":         "string",
+				"context_passed_ip":  "string",
+				"context_request_ip": "string",
+				"event_text":         "string",
+				"group_id":           "string",
+				"id":                 "string",
+				"loaded_at":          "datetime",
+				"name":               "string",
+				"original_timestamp": "datetime",
+				"previous_id":        "string",
+				"received_at":        "datetime",
+				"record_id":          "string",
+				"sent_at":            "datetime",
+				"timestamp":          "datetime",
+				"user_id":            "string",
+				"uuid_ts":            "datetime", // end of 18 rudder columns
+				"ad_name":            "string",
+				"blendo_id":          "string",
+			},
+		}
+		// With max col limit: 20
+		ExcludedSchema := warehouseutils.SchemaT{
+			"demo_event": {
+				"date_start":   "datatime",
+				"sample_event": "json",
+			},
+		}
+
+		It("Should not contain excluded schema if max column count is not reached", func() {
+			excludedSchema := GetExcludedSchema(uploadSchema, schemaInWarehouse, 1000)
+			Expect(excludedSchema).To(Equal(warehouseutils.SchemaT{}))
+		})
+
+		It("Should contain excluded schema if max column count is reached and picked in sorted order", func() {
+			excludedSchema := GetExcludedSchema(uploadSchema, schemaInWarehouse, 22)
+			Expect(excludedSchema).To(Equal(ExcludedSchema))
+		})
+
+		It("Should contain rudder reserved columns in included column list if it is a new event and max limit is reached", func() {
+			delete(schemaInWarehouse, "demo_event")
+			excludedSchema := GetExcludedSchema(uploadSchema, schemaInWarehouse, 22)
+			Expect(excludedSchema).To(Equal(ExcludedSchema))
+		})
+	})
+
+	Describe("Table schema diff", func() {
+		tableName := "demo_event"
+		uploadSchema := warehouseutils.SchemaT{
```

Comment:
Let's move this into the BeforeEach component instead of initializing it once.

Reply:
All initializations needs to be done in BeforeEach.

Reply:
If a map can be used globally, why not use it after initialising once instead of initializing before each test

Reply:
Any spec can update the map, so in order to have uniqueness we need to set it in BeforeEach which will ensure that each spec will get a new initialisation.